File: frontend/src/app/services/character.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../api-config';

export interface Character {
  id?: number;
  name: string;
  gender: string;
  hairStyle: string;
  hairColor: string;
  faceStyle: string;
  outfitStyle: string;
  title?: string;
  level?: number;
}

@Injectable({
  providedIn: 'root'
})
export class CharacterService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${API_BASE_URL}/api/characters`;

  saveCharacter(character: Character): Observable<any> {
    return this.http.post(this.baseUrl, character);
  }

  getMyCharacter(): Observable<Character> {
    return this.http.get<Character>(`${this.baseUrl}/me`);
  }
}

File: backend/src/main/java/com/englishascension/backend/feature/character/service/CharacterService.java
package com.englishascension.backend.feature.character.service;

import com.englishascension.backend.feature.character.dto.CharacterRequest;
import com.englishascension.backend.feature.user.entity.User;
import com.englishascension.backend.feature.user.repository.UserRepository;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.Map;

@Service
@Transactional
public class CharacterService {

    private final UserRepository userRepository;

    public CharacterService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public void saveCharacter(CharacterRequest characterRequest, String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found with email: " + email));

        user.setCharacterName(characterRequest.getName());
        user.setCharacterGender(characterRequest.getGender());
        user.setCharacterHairStyle(characterRequest.getHairStyle());
        user.setCharacterHairColor(characterRequest.getHairColor());
        user.setCharacterFaceStyle(characterRequest.getFaceStyle());
        user.setCharacterOutfitStyle(characterRequest.getOutfitStyle());
        if (user.getCharacterTitle() == null) {
            user.setCharacterTitle("Novice");
        }

        userRepository.save(user);
    }

    public Map<String, Object> getMyCharacter(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found with email: " + email));

        if (user.getCharacterName() == null) {
            return null;
        }

        Map<String, Object> character = new HashMap<>();
        character.put("id", user.getId());
        character.put("name", user.getCharacterName());
        character.put("gender", user.getCharacterGender());
        character.put("hairStyle", user.getCharacterHairStyle());
        character.put("hairColor", user.getCharacterHairColor());
        character.put("faceStyle", user.getCharacterFaceStyle());
        character.put("outfitStyle", user.getCharacterOutfitStyle());
        character.put("title", user.getCharacterTitle());
        character.put("level", user.getLevel());

        return character;
    }
}
